File: python-grpc-v2/python_grpc_v2/app/service/greeting_service.py
import uuid

from python_grpc_v2.app.adapter.publisher.greeting_publisher import GreetingPublisher
from python_grpc_v2.app.adapter.repo.greeting_repo import GreetingRepo
from python_grpc_v2.common.metaclass.singalton_meta import SingletonMeta
from python_grpc_v2.domain.entity.greeting import Greeting
from python_grpc_v2.infra.orm.connection import MySession


class GreetingService(metaclass=SingletonMeta):
    def __init__(self):
        self.__greeting_repo = GreetingRepo()
        self.__greeting_publisher = GreetingPublisher()

    async def save(self, user_name: str, sender: int, recipient: int, msg: str, db: MySession):
        greeting = Greeting(id=uuid.uuid4(), user_name=user_name, sender=sender, recipient=recipient, msg=msg)
        await self.__greeting_repo.save(greeting, db)
        await db.commit()

    async def send_grpc(self, user_name: str, sender: int, recipient: int, msg: str, db: MySession) -> str:
        greeting = Greeting(id=uuid.uuid4(), user_name=user_name, sender=sender, recipient=recipient, msg=msg)
        return await self.__greeting_publisher.publish(greeting)

File: java-grpc-v2/app/src/main/java/idv/sheng/java_grpc_v2/app/service/GreetingService.java
package idv.sheng.java_grpc_v2.app.service;

import idv.sheng.java_grpc_v2.app.dto.GreetingDto;
import idv.sheng.java_grpc_v2.app.ports.publisher.GreetingPublisher;
import idv.sheng.java_grpc_v2.app.ports.repo.GreetingRepo;
import idv.sheng.java_grpc_v2.domain.entity.Greeting;
import idv.sheng.java_grpc_v2.domain.valueobj.ServerType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.UUID;
import java.util.concurrent.TimeoutException;

@Service
public class GreetingService {
    @Autowired
    private GreetingPublisher greetingPublisher;
    @Autowired
    private GreetingRepo greetingRepo;

    public String greet(GreetingDto dto) throws InterruptedException, TimeoutException {
        return greetingPublisher.publish(dto);
    }

    public String beGreeted(GreetingDto dto) {
        Greeting greeting = Mapper.toEntity(dto);
        greetingRepo.save(greeting);
        return "Hello " + dto.userName();
    }

    static class Mapper {
        public static Greeting toEntity(GreetingDto dto) {
            return Greeting.builder()
                    .id(UUID.randomUUID())
                    .userName(dto.userName())
                    .sender(ServerType.of(dto.senderCode()))
                    .recipient(ServerType.of(dto.recipientCode()))
                    .msg(dto.msg())
                    .build();
        }
    }
}
